### ISSUE TITLE
big change identified in Attention module

### ISSUE THREAD

**Issue by drwangxian** [lucidrains/audiolm-pytorch#285]
Hello,  there, 

Regarding the module 'class Attention(nn.Module)' in audiolm_pytorch.py, I find you have made big change in this module, compared with standdard multihead attention. Specifically, if assume we are doing self attention, and the input x has shape (t, c), where t is the number of time steps, and c the number of channels. Assume we are using h heads, and the dimension for each head is d. 

1) For the query, 
q: x - > Linear -> (h, t, d)

2) For the key and value,
k/v: x -> Linear (t, d) -> repeat -> (h, t, d)

You use 1) to get a different query for each head. However, you use 2) to get a single k and v, which are shared for all the heads. This  is very different from common practice. Is my understanding correct? why do you do so?

**Comment by lucidrains** [comment]
hey Xian, that is a good catch

this project is a bit dated and I used to believe in an old paper https://arxiv.org/abs/1911.02150

however times have changed and it is now preferable to do grouped query attention. maybe we can leave this open and I'll modernize the practice and bump a minor version

**Comment by drwangxian** [comment]
lucidrains,

Appreciate your quick response!
The are tones of problems with the current implementation. Below I list some.
1) The  raw waves in a batch are always padded to the longest length allowed, rather than the longest length in the current batch.
2) The processing of different numbers of tokens in batch is rather vague. After tracing step-by-step, I find it is processed in the final loss function.
3) You are using hyper connection in place of standdard residual connection in the transformer encoder. However, you did not abide by the original paper, but instead for residual connection you just used the original direct sum.
4) There are problems in using EnCodec, in particular, the scaling problem.

I am transferring to pytorch from tensorflow. Because of my tf experience, I cannot stand with such loose style of implementation. 
With your help, next I am going to throughly fix potential problems and present a clean code. Hopefully, I can finally provide a trained model for others.

Please support me with your help. Thanks in advance!

**Comment by drwangxian** [comment]
Thanks for clarifying this! Now I understand your existing attention method is called multi-query attention, under which different heads use different quries, whereas keys and values are shared by all the heads. Moreover, grouped-query attention generalizes multi-query attention, resembling grouped convolution.
